### PR TITLE
fix(eventindexer): nft balance changes should be wrapped in db transaction

### PR DESCRIPTION
**Pull request by the-laziest**
Increase and subtract of nft balance in db should be done in transaction to prevent corrupted data in case of error in second query

### CODE REVIEW

**Review by xiaodino — DISMISSED** [review]
@the-laziest Could you explain more what the PR is for and a way to reproduce the issue?

**Comment by the-laziest** [comment]
> @the-laziest Could you explain more what the PR is for and a way to reproduce the issue?

Sure, @xiaodino,  if there will be an error for some reason [here](https://github.com/taikoxyz/taiko-mono/blob/215a6001c2c0dc0e8d2d20a438be9889792ff6ef/packages/eventindexer/indexer/index_nft_transfers.go#L135) (related to db connection or smth like that), then in database NFT balance will be increased for `To` address, but not decreased for `From` address

**Comment by cyberhorsey** [comment]
Two notes:

1) I'm happy to merge this PR, however, we will not use any of this code, and it will be deleted for mainnet (since we will use professional NFT indexing APIs that already exist), so it is not robust on purpose.

2) We can delete the existing `IncreaseBalance` and `SubtractBalance` methods now, since they are unused, so if this PR is to be merged, we should do that.

**Comment by the-laziest** [comment]
@cyberhorsey, updated all mentioned things

File: packages/eventindexer/indexer/index_nft_transfers.go [hunk 1]
```diff
@@ -116,35 +116,34 @@ func (i *Indexer) saveERC721Transfer(ctx context.Context, chainID *big.Int, vLog
...
 		Address:         to,
 		TokenID:         tokenID,
 		ContractAddress: vLog.Address.Hex(),
 		ContractType:    "ERC721",
 		Amount:          1, // ERC721 is always 1
-	})
-	if err != nil {
-		return err
 	}
+	descreaseOpts := eventindexer.UpdateNFTBalanceOpts{}
```
Comment (cyberhorsey) [inline packages/eventindexer/indexer/index_nft_transfers.go:128]: typo, should be decrease not descrease

File: packages/eventindexer/nft_balance.go [hunk 2]
```diff
@@ -30,8 +30,11 @@ type UpdateNFTBalanceOpts struct {
 
 // NFTBalanceRepository is used to interact with nft balances in the store
 type NFTBalanceRepository interface {
-	SubtractBalance(ctx context.Context, opts UpdateNFTBalanceOpts) (*NFTBalance, error)
-	IncreaseBalance(ctx context.Context, opts UpdateNFTBalanceOpts) (*NFTBalance, error)
+	IncreaseAndSubtractBalancesInTx(
+		ctx context.Context,
+		increaseOpts UpdateNFTBalanceOpts,
+		subtractOpts UpdateNFTBalanceOpts,
+	) (*NFTBalance, *NFTBalance, error)
```
Comment (cyberhorsey) [inline packages/eventindexer/nft_balance.go:37]: returning two pointers to the same strict here is not clear. we should either give them names return parameters, or wrap them in a parent struct and give each a named member, like "IncreasedNFTBalance" and "DecreasedNFTBalance" on the parent struct of "NFTBalances"

**Review by cyberhorsey — CHANGES_REQUESTED** [review]

**Comment by cyberhorsey** [comment]
Left two comments.

**Comment by the-laziest** [comment]
@cyberhorsey , could you please take a look on it?

**Review by cyberhorsey — APPROVED** [review]

**Comment by cyberhorsey** [comment]
LGTM. Thanks for doing the requested changes.

**Review by dantaik — APPROVED** [review]

**Comment by the-laziest** [comment]
@cyberhorsey , can I please ask for contributor poap, if it's applicable to this pr

**Comment by RogerLamTd** [comment]
@gitpoap-bot please give @the-laziest a poap

**Pull request merged**